This makes no sense...
```
FAILED: unittests/runtime/SwiftRuntimeTests.exe 
cmd.exe /C "cd . && "C:\Program Files\CMake\bin\cmake.exe" -E vs_link_exe --intdir=unittests\runtime\CMakeFiles\SwiftRuntimeTests.dir --rc=C:\PROGRA~2\WI3CF2~1\10\bin\100183~1.0\x64\rc.exe --mt=C:\PROGRA~2\WI3CF2~1\10\bin\100183~1.0\x64\mt.exe --manifests  -- C:\PROGRA~2\MICROS~1\2019\COMMUN~1\VC\Tools\MSVC\1424~1.283\bin\Hostx64\x64\link.exe /nologo @CMakeFiles\SwiftRuntimeTests.rsp  /out:unittests\runtime\SwiftRuntimeTests.exe /implib:unittests\runtime\SwiftRuntimeTests.lib /pdb:unittests\runtime\SwiftRuntimeTests.pdb /version:0.0  /INCREMENTAL:NO /STACK:10000000 /INCREMENTAL:NO /subsystem:console  && cd ."
LINK: command "C:\PROGRA~2\MICROS~1\2019\COMMUN~1\VC\Tools\MSVC\1424~1.283\bin\Hostx64\x64\link.exe /nologo @CMakeFiles\SwiftRuntimeTests.rsp /out:unittests\runtime\SwiftRuntimeTests.exe /implib:unittests\runtime\SwiftRuntimeTests.lib /pdb:unittests\runtime\SwiftRuntimeTests.pdb /version:0.0 /INCREMENTAL:NO /STACK:10000000 /INCREMENTAL:NO /subsystem:console /MANIFEST /MANIFESTFILE:unittests\runtime\SwiftRuntimeTests.exe.manifest" failed (exit code 1120) with the following output:
   Creating library unittests\runtime\SwiftRuntimeTests.lib and object unittests\runtime\SwiftRuntimeTests.exp
TaskFuture.cpp.obj : error LNK2019: unresolved external symbol "struct swift::FullMetadata<struct swift::TargetOpaqueMetadata<struct swift::InProcess> > const $sSiN" (?$sSiN@@3U?$FullMetadata@U?$TargetOpaqueMetadata@UInProcess@swift@@@swift@@@swift@@B) referenced in function "private: virtual void __cdecl TaskFutureTest_intFuture_Test::TestBody(void)" (?TestBody@TaskFutureTest_intFuture_Test@@EEAAXXZ)
unittests\runtime\SwiftRuntimeTests.exe : fatal error LNK1120: 1 unresolved externals
```

--- a/include/swift/ABI/Task.h
+++ b/include/swift/ABI/Task.h
@@ -31,6 +31,7 @@ class AsyncContext;
 class Executor;
 class Job;
 struct OpaqueValue;
+struct SwiftError;
 class TaskStatusRecord;
 
 /// An ExecutorRef isn't necessarily just a pointer to an executor
@@ -321,6 +322,12 @@ public:
     OpaqueValue *getStoragePtr() {
       return reinterpret_cast<OpaqueValue *>(
           reinterpret_cast<char *>(this) + storageOffset(resultType));
+    }
+
+    /// Retrieve the error.
+    SwiftError *&getError() {
+      return *reinterpret_cast<SwiftError **>(
+           reinterpret_cast<char *>(this) + storageOffset(resultType));
     }
 
     /// Compute the offset of the storage from the base of the future

--- a/include/swift/ABI/Task.h
+++ b/include/swift/ABI/Task.h
@@ -246,7 +246,7 @@ public:
   public:
     /// Describes the status of the future.
     ///
-    /// Futures always being in the "Executing" state, and will always
+    /// Futures always begin in the "Executing" state, and will always
     /// make a single state change to either Success or Error.
     enum class Status : uintptr_t {
       /// The future is executing or ready to execute. The storage
@@ -266,7 +266,6 @@ public:
     std::atomic<Status> status;
 
     /// Queue containing all of the tasks that are waiting in `get()`.
-    /// FIXME: do we also need a context pointer for each?
     std::atomic<AsyncTask*> waitQueue;
 
     /// The type of the result that will be produced by the future.
@@ -298,8 +297,6 @@ public:
           reinterpret_cast<char *>(this) + storageOffset(resultType));
     }
 
-    /// Retrieve a reference to the storage of the 
-
     /// Compute the offset of the storage from the base of the future
     /// fragment.
     static size_t storageOffset(const Metadata *resultType)  {
@@ -329,9 +326,8 @@ public:
   ///
   /// \returns the status of the future. If this result is
   /// \c Executing, then \c waitingTask has been added to the
-  /// wait queue and will be scheduled when the future completes or
-  /// is cancelled. Otherwise, the future has completed and can be
-  /// queried.
+  /// wait queue and will be scheduled when the future completes. Otherwise,
+  /// the future has completed and can be queried.
   FutureFragment::Status waitFuture(AsyncTask *waitingTask);
 
   /// Complete this future.

--- a/include/swift/ABI/Task.h
+++ b/include/swift/ABI/Task.h
@@ -88,6 +88,13 @@ public:
 
 /// A schedulable job.
 class alignas(2 * alignof(void*)) Job {
+protected:
+  // Indices into SchedulerPrivate, for use by the runtime.
+  enum {
+    /// The next waiting task link, an AsyncTask that is waiting on a future.
+    NextWaitingTaskIndex = 0,
+  };
+
 public:
   // Reserved for the use of the scheduler.
   void *SchedulerPrivate[2];
@@ -190,16 +197,12 @@ public:
   /// Reserved for the use of the task-local stack allocator.
   void *AllocatorPrivate[4];
 
-  /// The next task in the linked list of waiting tasks.
-  AsyncTask *NextWaitingTask;
-
   AsyncTask(const HeapMetadata *metadata, JobFlags flags,
             TaskContinuationFunction *run,
             AsyncContext *initialContext)
     : HeapObject(metadata), Job(flags, run),
       ResumeContext(initialContext),
-      Status(ActiveTaskStatus()),
-      NextWaitingTask(nullptr) {
+      Status(ActiveTaskStatus()) {
     assert(flags.isAsyncTask());
   }
 
@@ -265,9 +268,9 @@ public:
     /// head of the list of tasks.
     struct WaitQueueItem {
       /// Mask used for the low status bits in a wait queue item.
-      const uintptr_t statusMask = 0x03;
+      static const uintptr_t statusMask = 0x03;
 
-      uintptr_t storage;
+      const uintptr_t storage;
 
       Status getStatus() const {
         return static_cast<Status>(storage & statusMask);
@@ -359,13 +362,20 @@ public:
   /// executor.
   void completeFuture(AsyncContext *context, ExecutorRef executor);
 
+  /// Access the next waiting task, which establishes a singly linked list of
+  /// tasks that are waiting on a future.
+  AsyncTask *&getNextWaitingTask() {
+    return reinterpret_cast<AsyncTask *&>(
+        SchedulerPrivate[NextWaitingTaskIndex]);
+  }
+
   static bool classof(const Job *job) {
     return job->isAsyncTask();
   }
 };
 
 // The compiler will eventually assume these.
-static_assert(sizeof(AsyncTask) == 14 * sizeof(void*),
+static_assert(sizeof(AsyncTask) == 12 * sizeof(void*),
               "AsyncTask size is wrong");
 static_assert(alignof(AsyncTask) == 2 * alignof(void*),
               "AsyncTask alignment is wrong");

--- a/stdlib/public/Concurrency/Task.cpp
+++ b/stdlib/public/Concurrency/Task.cpp
@@ -34,7 +34,7 @@ void FutureFragment::destroy() {
     break;
 
   case Status::Error:
-    swift_unknownObjectRelease(getStoragePtr());
+    swift_unknownObjectRelease(reinterpret_cast<OpaqueValue *>(getError()));
     break;
   }
 }
@@ -300,8 +300,7 @@ swift::swift_task_future_wait(AsyncTask *task, AsyncTask *waitingTask) {
    case FutureFragment::Status::Error:
       return TaskFutureWaitResult{
           TaskFutureWaitResult::Error,
-          *reinterpret_cast<OpaqueValue **>(
-            task->futureFragment()->getStoragePtr())};
+          reinterpret_cast<OpaqueValue *>(task->futureFragment()->getError())};
   }
 }
 

--- a/stdlib/public/Concurrency/Task.cpp
+++ b/stdlib/public/Concurrency/Task.cpp
@@ -64,7 +64,7 @@ FutureFragment::Status AsyncTask::waitFuture(AsyncTask *waitingTask) {
     }
 
     // Put the waiting task at the beginning of the wait queue.
-    waitingTask->NextWaitingTask = queueHead.getTask();
+    waitingTask->getNextWaitingTask() = queueHead.getTask();
     auto newQueueHead = WaitQueueItem::get(Status::Executing, waitingTask);
     if (fragment->waitQueue.compare_exchange_weak(
             queueHead, newQueueHead, std::memory_order_release,
@@ -119,10 +119,10 @@ void AsyncTask::completeFuture(AsyncContext *context, ExecutorRef executor) {
   auto waitingTask = queueHead.getTask();
   while (waitingTask) {
     // Find the next waiting task.
-    auto nextWaitingTask = waitingTask->NextWaitingTask;
+    auto nextWaitingTask = waitingTask->getNextWaitingTask();
 
     // Remove this task from the list.
-    waitingTask->NextWaitingTask = nullptr;
+    waitingTask->getNextWaitingTask() = nullptr;
 
     // TODO: schedule this task on the executor rather than running it
     // directly.

--- a/unittests/runtime/TaskFuture.cpp
+++ b/unittests/runtime/TaskFuture.cpp
@@ -196,7 +196,7 @@ TEST(TaskFutureTest, objectFuture) {
     EXPECT_EQ(object, *reinterpret_cast<TestObject **>(waitResult.storage));
 
     // Make sure the object hasn't been destroyed.
-    EXPECT_EQ(7, objectValueOnComplete);
+    EXPECT_EQ(size_t(7), objectValueOnComplete);
 
     // Okay, release the task. This should destroy the object.
     swift_release(task);

